refactor: link user to member (LAN-353 / LANDA-635)

barredterra opened the pull request:
### Motivation

Currently, **LANDA Member** is linked to **User** and **User** is linked to **LANDA Member**. **User** has to be created/enabled/disabled by modifying **LANDA Member**. This doesn't make sense.

This PR separates them more clearly. Now you can create a **LANDA Member** (independent of **User**), and, optionally, a **User** that is linked to the **LANDA Member**. When you want to create/enable/disable/delete the **User**, you do it directly in the **User** DocType, as one would expect.

### Changes

**LANDA Member**:

- Remove "LANDA Account" section
- Add button "Create User"

    Shown only if the a user for this member doesn't exist yet and the current user has permission to create new users.

- Add **User** to connections

**User**:

- Allow to toggle "Enabled" checkbox for anyone who can modify users (not just System Managers)
- Once a **LANDA Member** is linked, the field is set to _read only_
- Allow RWCD for roles "LANDA State Organization Management", "LANDA Regional Organization Management" and "LANDA Local Organization Management"

    Can reset passwords, and enable/disable/delete users

- Allow Select for "LANDA Member", disallow for "All".
- Set permlevel and defaults for fields that should not be edited by organization management
- Move custom fields, property setters and permissions into JSON file

**Member Function Category**:

- Get any _active_ **User** linked _to_ **LANDA Member**, instead of the **User** linked _in_ **LANDA Member**.

### What I tested

As a user with Role "LANDA Regional Organization Management", I

- created a new **User** for a member, checked for appropriate **User Permissions**
- disabled and re-enabled the **User**
- created an active **Member Function** and checked the **User** for appropriate roles
- disabled the **Member Function** and checked the **User** for appropriate roles

As a user with Role "LANDA Local Organization Management", I

- created a new User

### Notes

Before testing, be sure to run `bench migrate`, restart, `bench clear-cache` and refresh your browser window with `Shift + Cmd + R`.

barredterra committed:
refactor: link user to member

barredterra committed:
fix: remove unused imports

barredterra committed:
fix: always add "LANDA Member" to roles

barredterra committed:
feat: validate that only one active user is linked to a member

barredterra committed:
Merge branch 'version-13-hotfix' into member-user-link

melanieroedel commented:
Looking good so far. I like the new UX with the "Create User" button.
I started testing as an Administrator.

melanieroedel commented on landa/organization_management/doctype/landa_member/landa_member.json:
```diff
@@ -293,9 +254,14 @@
    "link_fieldname": "member"
+  },
+  {
+   "custom": 1,
```
```suggestion
```

Not sure what this does, but it leads to a double entry in connections for me. 
![image](https://user-images.githubusercontent.com/79863208/204783142-d952df02-b3b3-443e-b8ed-65687ac36fae.png)
Removing this line reduces it to one link in the connections that works just fine.

melanieroedel commented:
When a user with an existing member function is created the permissions are not set based on the existing member function, but to LANDA Member only.
Maybe call `on_update` from a new function `on_create` (if that exists) in `landa/organization_management/user/user.py`?

melanieroedel commented:
We should change the defaults and the permission level of these two sections:
![image](https://user-images.githubusercontent.com/79863208/204786876-a252aeb0-ded5-4b2e-be79-ecde44f3e852.png)
This is what is the view wen creating a user as a local organisation:
<img width="1341" alt="image" src="https://user-images.githubusercontent.com/79863208/204789324-051f5937-8df8-4dc9-878c-8f7a6ee9baee.png">

My suggestion is setting the defaults to 0 for all checkboxes and changing the permission level to 1 to make sure local organisation users can't change it.

Users should not be allowed in mentions, so the default of "Allow in Mentions" has to be 0. If any, only users who are employees should be allowed to be mentioned- which can be set manually by an administrator. This could be an issue because technically users can change that for their own user. Any ideas for preventing this? It seems awfully complicated to only allow mentions for users that have a role on the regional level.

Not sure what "Send Notifications For Email Threads" and "Send Me A Copy of Outgoing Emails" do, but they probably should be set to 0 by default. 
"Send Notifications For Documents Followed By Me" could be useful, I guess, but it's better to leave it at 0 by default.

melanieroedel commented:
Currently users can change the member and thus the permissions when the wrongly linked member's function is changed.
The field `landa_member` could be changed to `"set_only_once": 1`.
Now that regional organisation users can delete a user, they can just delete and start over if there was a mix up.

melanieroedel commented:
Local organisation users cannot create users because they have no access to enter the field email:
<img width="1389" alt="image" src="https://user-images.githubusercontent.com/79863208/204799895-ccb674ca-deac-400e-9296-4e6503b654aa.png">
One workaround could be to fetch the primary email address from one of the member's contacts.

barredterra committed:
Merge remote-tracking branch 'upstream/version-13-hotfix' into member-user-link

barredterra committed:
fix: permission check (LAN-353)

barredterra committed:
fix: trigger permissions after insert of User (LAN-353)

barredterra committed:
fix: user connection in member (LAN-353)

barredterra committed:
feat: default values for User (LAN-353)

Disable check boxes document_follow_notify, allowed_in_mentions, thread_notify by default.

barredterra commented:
- Fixed permission for showing "Create User" button (currently not allowed for local organization)
- Fixed duplicate User connection (good catch!)
- Set permissions when creating a new user
- Disabled checkboxes by default

melanieroedel commented:
Now local organisation users have neither the plus behind "User" in Connections nor the "Create User" button on the top right.
Regional organisation users have the plus behind "User" in Connection, but not the "Create User" button on the top right.
Neither has access to the list view of "User" (the DocType).
Both for regional and local users I get an error 
![image](https://user-images.githubusercontent.com/79863208/205329928-8266f795-8e1b-4931-ba5c-9468aaedf6bb.png)
when trying to access the DocType User (either via list view or via Connections)
It works after clearing my cache and reloading for each user individually though. Very strange...

barredterra commented:
> It works after clearing my cache and reloading for each user individually though.

Cache problems don't count 😁

melanieroedel commented:
> > It works after clearing my cache and reloading for each user individually though.
> 
> Cache problems don't count 😁

It is very repeatable though. Even for the _same user_ (if I log in as another one and log back in with the original user later)  🤔

melanieroedel commented:
> Now local organisation users have neither the plus behind "User" in Connections nor the "Create User" button on the top right.
> Regional organisation users have the plus behind "User" in Connection, but not the "Create User" button on the top right.

This issue is independent of the cache problem

barredterra committed:
Merge branch 'version-13-hotfix' into member-user-link

barredterra committed:
feat: set only once for link to LANDA Member

barredterra committed:
fix: role permissions on user

barredterra committed:
fix: allow form_sidebar for LANDA Local Organization Management

Required for seeing menu and custom buttons

barredterra committed:
fix: permlevel and default for some fields in user

Permlevel 1 on:
- document_follow_notify
- logout_all_sessions
- new_password
- send_me_a_copy
- email_signature
- allowed_in_mentions
- thread_notify
- short_bio

Defaults for:
- time_zone
- document_follow_notify
- allowed_in_mentions
- thread_notify

melanieroedel commented:
`bench migrate` doesn't run through. Probably an issue in `/home/frappe/bench-landa/apps/landa/landa/install.py` when setting `time_zone`.

```
Migrating landa.local
Updating DocTypes for frappe        : [========================================] 100%
Updating DocTypes for erpnext       : [========================================] 100%
Updating DocTypes for landa         : [========================================] 100%
Updating Dashboard for frappe
Updating Dashboard for erpnext
Updating Dashboard for landa
Updating customizations for Address
Updating customizations for Contact
Updating customizations for User

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/frappe/bench-landa/apps/frappe/frappe/utils/bench_helper.py", line 110, in <module>
    main()
  File "/home/frappe/bench-landa/apps/frappe/frappe/utils/bench_helper.py", line 20, in main
    click.Group(commands=commands)(prog_name="bench")
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/frappe/bench-landa/env/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/frappe/bench-landa/apps/frappe/frappe/commands/__init__.py", line 31, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/frappe/bench-landa/apps/frappe/frappe/commands/site.py", line 433, in migrate
    migrate(context.verbose, skip_failing=skip_failing, skip_search_index=skip_search_index)
  File "/home/frappe/bench-landa/apps/frappe/frappe/migrate.py", line 81, in migrate
    sync_customizations()
  File "/home/frappe/bench-landa/apps/frappe/frappe/modules/utils.py", line 112, in sync_customizations
    sync_customizations_for_doctype(data, folder)
  File "/home/frappe/bench-landa/apps/frappe/frappe/modules/utils.py", line 171, in sync_customizations_for_doctype
    validate_fields_for_doctype(doctype)
  File "/home/frappe/bench-landa/apps/frappe/frappe/core/doctype/doctype/doctype.py", line 907, in validate_fields_for_doctype
    validate_fields(meta)
  File "/home/frappe/bench-landa/apps/frappe/frappe/core/doctype/doctype/doctype.py", line 1315, in validate_fields
    check_illegal_default(d)
  File "/home/frappe/bench-landa/apps/frappe/frappe/core/doctype/doctype/doctype.py", line 1052, in check_illegal_default
    frappe.throw(
  File "/home/frappe/bench-landa/apps/frappe/frappe/__init__.py", line 504, in throw
    msgprint(
  File "/home/frappe/bench-landa/apps/frappe/frappe/__init__.py", line 479, in msgprint
    _raise_exception()
  File "/home/frappe/bench-landa/apps/frappe/frappe/__init__.py", line 434, in _raise_exception
    raise raise_exception(msg)
frappe.exceptions.ValidationError: Options for <b>time_zone</b> must be set before setting the default value.
```

barredterra committed:
fix: remove property setter for time zone

barredterra commented:
@melanieroedel removed this property setter, as it's not important

melanieroedel committed:
feat: add user related translations

melanieroedel commented:
Nice one larger request and a couple of smaller ones.
Important: Local organization users should have the same permissions as regional organization users for this PR.

melanieroedel commented:
If I go to the user lis, add a user from scratch and leave the landa-member field empty, the code throws an error (as it should). But this error message may be a bit confusing to the user:
<img width="535" alt="image" src="https://user-images.githubusercontent.com/79863208/205977661-66292179-603d-4809-8007-5c651d05720f.png">
Not a big deal. We could also just leave it as is.
The error message should be 
`Not allowed for {0}: {1}. Restricted field: {2},`
but I couldn't find the origin in frappe.

melanieroedel commented:
Deletes are not possible for active users, but that should be covered by the new delete functionality anyway.
There is a translation issue with regard to deletes. The DocTypes in the error message do not get translated:
`Kann nicht gelöscht oder abgebrochen werden, weil User nutzer@test.de mit Route History 7291e02257 verknüpft ist`
which should be
`Cannot delete or cancel because {0} {1} is linked with {2} {3} {4}`
Not sure if that can be easily fixed in frappe.

barredterra commented:
> The DocTypes in the error message do not get translated

https://github.com/frappe/frappe/pull/19164

> this error message may be a bit confusing to the user

https://github.com/frappe/frappe/pull/19165

> Important: Local organization users should have the same permissions as regional organization users for this PR.

This should already be the case.

barredterra committed:
Merge branch 'version-13-hotfix' into member-user-link

melanieroedel reviewed (approved):
After a
`bench setup requirements && bench build --force && bench migrate && bench restart && bench clear-cache` it's now all performing as specified. Sorry about the confusion. Well done!

barredterra commented:
:tada: This PR is included in version 0.2.0 :tada:

The release is available on [GitHub release](https://github.com/alyf-de/landa/releases/tag/v0.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
